Source/engine/random.hpp

```diff
@@ -40,7 +40,18 @@ void DiscardRandomValues(unsigned count);
 /**
  * @brief Advances the global RandomNumberEngine state and returns the new value
  */
-uint32_t GenerateSeed();
+[[nodiscard]] uint32_t GenerateSeed();
+
+/**
+ * @brief Provides the seed value for a sequence of random numbers of predetermined size
+ *
+ * This advances the engine state of the random number generator by count+1 rounds to ensure that
+ * subsequently generated random numbers will not overlap with the allocated sequence.
+ *
+ * @param count The size of the random number sequence to be allocated
+ * @return The seed for the allocated sequence of random numbers
+ */
+[[nodiscard]] uint32_t AllocateRandomSequence(unsigned count);
```

Comment:
You've mentioned what this method is intended for (ensuring later uses of RNG state don't have immediately overlapping sequences) however I think this could be made clearer and "safer" by removing the ability to pass a parameter here or potentially even using a different approach entirely.

To fit the way C++ random number concepts are defined we should use a completely different random number generator to generate seeds later intended for use with the Borland LCG engine. I was thinking of proposing [xoshiro128++](https://prng.di.unimi.it) as the global RNG since it requires far less state than a Mersenne Twister (the idea was this would become the backing generator for `GenerateSeed()` and calls to `SetRndSeed()` would become local instances of `diabloGenerator`s).

That said, this approach is effectively the same as xoshiro/xoroshiro `jump()` or [java.util.SplittableRandom.split()](https://docs.oracle.com/javase/8/docs/api/java/util/SplittableRandom.html#split--) so is perfectly serviceable as is.

Reply:
I don't know how you would allocate RNG space for a sequence of random numbers with predetermined length if you don't take that predetermined length as a parameter. But I guess what you're saying is using a different RNG algorithm to generate seeds for the LCG would place us at a random point in the sequence, mitigating overlaps without the need for predetermination of the size of the RNG sequence?

Reply:
Yeah, the main thing I was hoping for would be to provide suitable default behaviour that call sites don't need to determine the appropriate length for their use case. For this use you've noted that in the best case somewhere between 2^11 and 2^13 values are needed, but the worst case would require a gap of 2^19 or more to avoid all potential overlaps to fully resolve the linked issue. Using a different engine to seed this engine doesn't fully solve the issue either since there would still be a chance that two consecutive outputs from the first engine land near each other in the second engines cycle, but it's a lot less to think about (and works well for engines that don't have the ability to jump multiple values ahead).

jump/split functions in other implementations effectively partition their state space into chunks of a given size (e.g. 2^64 non-overlapping chunks of 2^64 values for xoshiro128), maybe 2^15 is a reasonable default here and you can let callers use a smaller jump length if they know it's safe for their purposes.

Reply:
> ... but the worst case would require a gap of 2^19 or more to avoid all potential overlaps to fully resolve the linked issue.

That's not quite right. `LoadGameLevel()` reseeds the RNG four times using the same seed, essentially splitting the process into four phases that all share the same RNG sequence. The linked issue is concerned with item generation, which is dependent on the object and monster seeds generated in phase 4. The worst case for phase 4 was measured around 6000 values, which I believe is what you were referring to with 2^13.

The part that might require 2^19 values is phase 3, which is the part that generates the dungeon layout and is therefore largely unrelated to the linked issue. I don't believe it's critical to avoid overlaps in phase 3. But if you really wanted to, you might be able to get away with reducing the random value requirement by taking a more conservative approach than discarding a whole chunk of the RNG sequence each time the level generation fails.

> Using a different engine to seed this engine doesn't fully solve the issue either...

The solution I've implemented here should fully solve the issue, and it is true that your solution can't provide the same guarantee that there would be no overlap. However, I think your solution would be good enough in that it would provide reasonable assurance that more often than not there should be no significant overlap. So I have no objections.

Reply:
> > ... but the worst case would require a gap of 2^19 or more to avoid all potential overlaps to fully resolve the linked issue.
> 
> That's not quite right. `LoadGameLevel()` reseeds the RNG four times using the same seed, essentially splitting the process into four phases that all share the same RNG sequence. The linked issue is concerned with item generation, which is dependent on the object and monster seeds generated in phase 4. The worst case for phase 4 was measured around 6000 values, which I believe is what you were referring to with 2^13.

Yeah I was using 2^13 as an approximation of ~6000 rolls, it was easier for me to think of in terms of bits since that leads into chunking the engine period/cycle. Also I completely forgot about the constant re-seeding of the global state during level load... That makes it easier to be confident a shorter stride is enough to avoid repetitions between levels for the target area.